server/rosetta/cosmos/conversion/to_rosetta.go

```diff
@@ -0,0 +1,275 @@
+	}
+
+	for i, coin := range availableCoins {
+		value, owned := ownedCoinsMap[coin.Denom]
+		if !owned {
+			amounts[i] = &types.Amount{
+				Value: sdk.NewInt(0).String(),
+				Currency: &types.Currency{
+					Symbol: coin.Denom,
+				},
+			}
+			continue
+		}
+		amounts[i] = &types.Amount{
+			Value: value.String(),
+			Currency: &types.Currency{
+				Symbol: coin.Denom,
+			},
+		}
+	}
+
+	return amounts
+}
+
+// SdkTxsWithHashToRosettaTxs converts sdk transactions wrapped with their hash to rosetta transactions
+func SdkTxsWithHashToRosettaTxs(txs []*rosetta.SdkTxWithHash) []*types.Transaction {
+	converted := make([]*types.Transaction, len(txs))
+	for i, tx := range txs {
+		hasError := tx.Code != 0
+		converted[i] = &types.Transaction{
+			TransactionIdentifier: &types.TransactionIdentifier{Hash: tx.HexHash},
+			Operations:            SdkTxToOperations(tx.Tx, true, hasError),
+			Metadata: map[string]interface{}{
+				rosetta.Log: tx.Log,
+			},
+		}
+	}
+
+	return converted
+}
+
+// SdkTxToOperations converts an sdk.Tx to rosetta operations
+func SdkTxToOperations(tx sdk.Tx, withStatus, hasError bool) []*types.Operation {
+	var operations []*types.Operation
+
+	feeTx := tx.(sdk.FeeTx)
+	feeOps := sdkFeeTxToOperations(feeTx, withStatus)
+	operations = append(operations, feeOps...)
+
+	msgOps := sdkMsgsToRosettaOperations(tx.GetMsgs(), withStatus, hasError, len(feeOps))
+	operations = append(operations, msgOps...)
+
+	return operations
+}
+
+// sdkFeeTxToOperations converts sdk.FeeTx to rosetta operations
+func sdkFeeTxToOperations(feeTx sdk.FeeTx, withStatus bool) []*types.Operation {
+	feeCoins := feeTx.GetFee()
+	var ops []*types.Operation
+	if feeCoins != nil {
+		var feeOps = rosettaFeeOperationsFromCoins(feeCoins, feeTx.FeePayer().String(), withStatus)
+		ops = append(ops, feeOps...)
+	}
+
+	return ops
+}
+
+// rosettaFeeOperationsFromCoins returns the list of rosetta fee operations given sdk coins
+func rosettaFeeOperationsFromCoins(coins sdk.Coins, account string, withStatus bool) []*types.Operation {
+	feeOps := make([]*types.Operation, 0)
+	var status string
+	if withStatus {
+		status = rosetta.StatusSuccess
+	}
+
+	for i, coin := range coins {
+		op := &types.Operation{
+			OperationIdentifier: &types.OperationIdentifier{
+				Index: int64(i),
+			},
+			Type:   rosetta.OperationFee,
+			Status: status,
+			Account: &types.AccountIdentifier{
+				Address: account,
+			},
+			Amount: &types.Amount{
+				Value: "-" + coin.Amount.String(),
+				Currency: &types.Currency{
+					Symbol: coin.Denom,
+				},
+			},
+		}
+
+		feeOps = append(feeOps, op)
+	}
+
+	return feeOps
+}
+
+// sdkMsgsToRosettaOperations converts sdk messages to rosetta operations
+func sdkMsgsToRosettaOperations(msgs []sdk.Msg, withStatus bool, hasError bool, feeLen int) []*types.Operation {
+	var operations []*types.Operation
+	var status string
+	for i, msg := range msgs {
+		switch msg.Type() {
+		case rosetta.OperationSend:
```

Comment:
this is not scalable, can't we rather add a `RosettaOperation` function to each message type?